### PR TITLE
fix(package-linker): Fix issue with .bin folder with --modules-folder flag

### LINKED ISSUE #3724: custom node_modules folder breaks .bin since yarn 0.25.4
In our setting we install all our yarn packages in a different folder then the default node_modules folder with the --modules-folder setting.
e.g.:
```
yarn --modules-folder vendor/node_modules
```

**What is the current behavior?**
The current behaviour (version 0.24.*)  will install them in vendor/node_modules/.bin
After upgrading to version 0.25.4 it installs them incorrectly in node_modules/.bin, the packages are still installed in vendor/node_modules and the symlinks in node_modules/.bin still work.
However it seems some of them are missing, as I can't find uglifyjs binary for example in my project in that folder.

**What is the expected behavior?**
Install the packages as version 0.24.* did by installing them in vendor/node_modules/.bin

node.js v7.10.0
yarn: 0.24.5/0.25.4

### PR DESCRIPTION
**Pull request by voondo**
**Summary**

This PR tries to fix #3724 and #4297 where `install` was not installing binary symlinks to the correct location
when the `--modules-folder` flag was provided.

**Test plan**

Manual CLI tests:

Before:
```
> yarn --modules-folder /var/foo
> (ls /var/foo/.bin 2>1 > /dev/null && echo "found") || echo "not found"
not found
```

After:
```
> yarn --modules-folder /var/foo
> (ls /var/foo/.bin 2>1 > /dev/null && echo "found") || echo "not found"
found
```

### CODE REVIEW

File: src/package-linker.js [hunk 1]
```diff
@@ -436,7 +436,12 @@ export default class PackageLinker {
         topLevelDependencies,
         async ([dest, {pkg}]) => {
           if (pkg._reference && pkg._reference.location && pkg.bin && Object.keys(pkg.bin).length) {
-            const binLoc = path.join(this.config.lockfileFolder, this.config.getFolder(pkg));
+            let binLoc;
```
Comment (BYK) [inline src/package-linker.js]: Prefer `const`:
```js
const binLoc = this.config.modulesFolder ? this.config.modulesFolder : path.join(this.config.lockfileFolder, this.config.getFolder(pkg));
```

Comment (voondo) [reply src/package-linker.js]: done

Comment (voondo) [reply src/package-linker.js]: eslint is unhappy because `Line 439 exceeds the maximum line length of 120`. How should I write this line without `if` and `let` ?

**Comment by voondo** [comment]
Ok here is the latest state of my work but I encountered so many issues with `--modules-folder` I had to give up.

It contains a test as requested but I can't make it work (altough the manual CLI test works).
It also fixes a problem with `yarn --modules-folder /var/foo run [...]` (need also a test).
There are still some issues with `yarn check --modules-folder /var/foo --verify-tree`

TL;DR `--modules-folder` is currently too broken with yarn even with this patch, don't use it.

**Comment by BYK** [comment]
@voondo how can we help you to get this merged?

**Comment by voondo** [comment]
@BYK honestly I don't know. After digging a bit, I seen so many problems with this option.
By the way it has almost no test in the current code base (which could explain why it is broken).
Indeed, the only related test is `uses the modulesFolder configuration option as the first part of the path` in `__tests__/package-hoister.js`. From what I understand, it is far from sufficient.

On the other hand, in `src/`, we have dozens of occurrences of `path.join(cwd, 'node_modules'`, `path.join(config.cwd, registry.folder)`, `path.join(this.config.lockfileFolder, constants.NODE_MODULES_FOLDER)` or diferents variations which seems to refer to the same thing.

Therefore, the amount of work seems huge (especially for me who do not know well the secrets of yarn). 

Feel free to close this pull request but maybe it could help someone more courageous than me to get the work done ;)

**Comment by arcanis** [comment]
Yeah, `--modules-folder` is kind of a relicate from the time where Yarn was trying to be bower-compatible, iirc, so it's not as battle-proof as we'd like. I sometime wonder whether there's value in keeping it 😕

**Comment by kilianc** [comment]
@aracarie without `--modules-folder` docker workflows with mounted local volumes are broken. You mount your local dir that sometimes contains `node_modules` (standardjs, or other deps used by your IDE) and you want to use a different path in your container to avoid conflict with compiled deps.

The alternative is to `mv node_modules` after running `yarn` which is pretty hacky.

**Comment by voondo** [comment]
The use with docker was also my use case. I finally "solved" it by automatically running yarn when the container is starting in development. But I agree it is hacky and would be better to have a working `--modules-foler`.
@arcanis Until it is fixed, maybe you could disable this option to not mislead us (or at least a warning)

**Comment by kilianc** [comment]
Aside from being slower to re-run `yarn` at every cold start of the container it leads to all sort of side-effects in the opposite way, where now you have on your machine the deps used by the container.

At this moment, this is the best monkey-patch I found:

```dockerfile
RUN yarn --no-emoji --pure-lockfile

# https://github.com/yarnpkg/yarn/issues/4297
# https://github.com/yarnpkg/yarn/pull/5741
RUN mv ./node_modules $NODE_PATH
```
